File: max-consecutive-ones-iii.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    int longestOnes(vector<int>& A, int K) {
        int i = 0,j = 0;
        int ans = 0;

        while(j<A.size()){
            
            if(A[j] == 0){
                K--;
                if(K<0){
                    while(K<0){
                        if(A[i] == 0) K++;
                        i++;
                    }
                }
            }
            if(K>=0){
                ans = max(ans, j-i+1);
            }
            j++;
            
        }
        
        return ans;
        
    }
};

File: Max Consecutive Ones III.py
#time: O(n)
#space: O(1)
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        left = 0
        length = 0
        count = 0
        
        for right in range(len(nums)):
            if nums[right] == 1:
                count += 1
            
            while right-left+1 - count > k:
                if nums[left] == 1:
                    count -= 1
                left +=1
        
            length = max(length, right-left+1)
        
        return length